src/core/tsi/s2a/handshaker/s2a_tsi_handshaker.cc

```diff
@@ -16,34 +16,327 @@
+  if (self == nullptr || bytes == nullptr || bytes_size == nullptr) {
+    gpr_log(GPR_ERROR, kS2ATsiHandshakerResultUnusedBytesNullptr);
+    return TSI_INVALID_ARGUMENT;
+  }
+  s2a_tsi_handshaker_result* result =
+      reinterpret_cast<s2a_tsi_handshaker_result*>(
+          const_cast<tsi_handshaker_result*>(self));
+  *bytes = result->unused_bytes;
+  *bytes_size = result->unused_bytes_size;
+  return TSI_OK;
+}
+
+static void s2a_handshaker_result_destroy(tsi_handshaker_result* self) {
+  if (self == nullptr) {
+    return;
+  }
+  s2a_tsi_handshaker_result* result =
+      reinterpret_cast<s2a_tsi_handshaker_result*>(self);
+  // TODO(mattstev): do I need to add nullptr checks?
+  gpr_free(result->in_traffic_secret);
+  gpr_free(result->out_traffic_secret);
+  gpr_free(result->spiffe_id);
+  gpr_free(result->hostname);
+  gpr_free(result->unused_bytes);
+  gpr_free(result);
+}
+
+static const tsi_handshaker_result_vtable s2a_result_vtable = {
+    s2a_handshaker_result_extract_peer,
+    s2a_handshaker_result_create_zero_copy_grpc_protector, nullptr,
+    s2a_handshaker_result_get_unused_bytes, s2a_handshaker_result_destroy};
+
+// TODO(mattstev): I don't use the application protocol and local identity
+// fields from |response|, or the attributes field from the peer identity.
 tsi_result s2a_tsi_handshaker_result_create(s2a_SessionResp* response,
                                             bool is_client,
-                                            tsi_handshaker_result** result) {
-  return TSI_UNIMPLEMENTED;
+                                            tsi_handshaker_result** self) {
+  if (self == nullptr || response == nullptr) {
+    gpr_log(GPR_ERROR, kS2ATsiHandshakerResultNullptrArguments);
+    return TSI_INVALID_ARGUMENT;
+  }
+  const s2a_SessionResult* handshake_result = s2a_SessionResp_result(response);
+  if (handshake_result == nullptr) {
+    gpr_log(GPR_ERROR, kS2ATsiHandshakerResultEmpty);
+    return TSI_FAILED_PRECONDITION;
+  }
+  const s2a_Identity* peer_identity =
+      s2a_SessionResult_peer_identity(handshake_result);
+  if (peer_identity == nullptr) {
+    gpr_log(GPR_ERROR, kS2ATsiHandshakerResultInvalidPeerIdentity);
+    return TSI_FAILED_PRECONDITION;
+  }
+  const s2a_SessionState* handshake_state =
+      s2a_SessionResult_state(handshake_result);
+  if (handshake_state == nullptr) {
+    gpr_log(GPR_ERROR, kS2ATsiHandshakerResultInvalidSessionState);
+    return TSI_FAILED_PRECONDITION;
+  }
+  upb_strview in_traffic_secret = s2a_SessionState_in_key(handshake_state);
+  upb_strview out_traffic_secret = s2a_SessionState_out_key(handshake_state);
+  if (in_traffic_secret.size != out_traffic_secret.size) {
+    gpr_log(GPR_ERROR, kS2ATrafficSecretSizeMismatch);
+    return TSI_FAILED_PRECONDITION;
+  }
+
+  /** Instantiate S2A TSI handshaker result. **/
+  s2a_tsi_handshaker_result* tsi_result =
+      static_cast<s2a_tsi_handshaker_result*>(
+          gpr_malloc(sizeof(s2a_tsi_handshaker_result)));
```

Comment:
If we use `gpr_zalloc`, we can avoid some of `nullptr` assignments below.

Reply:
Done!